e2e_playwright/st_cache_resource.py

```diff
@@ -0,0 +1,70 @@
+# Copyright (c) Streamlit Inc. (2018-2022) Snowflake Inc. (2022-2024)
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import streamlit as st
+
+st.button("click to rerun")
+
+side_effects = []
+
+
+@st.cache_resource(experimental_allow_widgets=True)
+def foo():
+    side_effects.append("function ran")
+    r = st.radio("radio", ["foo", "bar", "baz", "qux"], index=1)
+    return r
+
+
+foo()
+
+st.text(side_effects)
+
+
+@st.cache_resource
+def with_cached_widget_warning():
+    st.write("Cached function that should show a widget usage warning.")
+    st.selectbox("selectbox", ["foo", "bar", "baz", "qux"], index=1)
+
+
+if st.button("Run cached function with widget warning"):
+    with_cached_widget_warning()
+
+
+@st.cache_resource(experimental_allow_widgets=True)
+def inner_cache_function():
+    st.radio("radio 2", ["foo", "bar", "baz", "qux"], index=1)
+
+
+@st.cache_resource(experimental_allow_widgets=False)
+def nested_cached_function():
+    inner_cache_function()
+    st.selectbox("selectbox 2", ["foo", "bar", "baz", "qux"], index=1)
+
+
+if st.button("Run nested cached function with widget warning"):
+    nested_cached_function()
```

Comment:
great test! Could you leave a comment here what the expectation is? E.g. 
When running `nested_cached_function()`, we get two warnings, one from `nested_cached_function()` and one from `inner_cache_function`. `inner_cache_function()` on its own would allow the usage, but since it is nested in the other function that does not allow it, we don't allow it. The reason for this is...

Reply:
Added a comment